### ISSUE TITLE
代码请求

### ISSUE THREAD

**Issue by akang-ai** [LiamLian0727/WaterMask#10]
import argparse
import os
import os.path as osp
import time
import warnings

import mmcv
import torch
from mmcv import Config
from mmcv.runner import load_checkpoint

from mmdet.apis import single_gpu_test
from mmdet.datasets import (build_dataloader, build_dataset,
                            replace_ImageToTensor)
from mmdet.models import build_detector
from mmdet.utils import (build_dp, get_device,
                         replace_cfg_vals, setup_multi_processes,
                         update_data_root)

def parse_args():
    parser = argparse.ArgumentParser(
        description='MMDet segmentation and save results')
    parser.add_argument('config', help='test config file path')
    parser.add_argument('checkpoint', help='checkpoint file')
    parser.add_argument(
        '--show-dir', help='directory where segmentation images will be saved')
    parser.add_argument(
        '--gpu-id',
        type=int,
        default=0,
        help='id of gpu to use '
        '(only applicable to non-distributed testing)')
    parser.add_argument(
        '--show-score-thr',
        type=float,
        default=0.3,
        help='score threshold (default: 0.3)')
    args = parser.parse_args()
    return args

def main():
    args = parse_args()

    with open(args.config, 'r', encoding='utf-8') as f:
        config_str = f.read()
    cfg_dict, cfg_text = Config._file2dict(args.config)

    cfg = Config(cfg_dict, cfg_text=cfg_text)

#    cfg = Config.fromfile(args.config)

    # 将 ${key} 替换为 cfg.key 的值
    cfg = replace_cfg_vals(cfg)

    # 根据 MMDET_DATASETS 更新数据根
    update_data_root(cfg)

    # 设置多线程设置
    setup_multi_processes(cfg)

    # 设置 cudnn_benchmark
    if cfg.get('cudnn_benchmark', False):
        torch.backends.cudnn.benchmark = True

    if 'pretrained' in cfg.model:
        cfg.model.pretrained = None
    elif 'init_cfg' in cfg.model.backbone:
        cfg.model.backbone.init_cfg = None

    if cfg.model.get('neck'):
        if isinstance(cfg.model.neck, list):
            for neck_cfg in cfg.model.neck:
                if neck_cfg.get('rfp_backbone'):
                    if neck_cfg.rfp_backbone.get('pretrained'):
                        neck_cfg.rfp_backbone.pretrained = None
        elif cfg.model.neck.get('rfp_backbone'):
            if cfg.model.neck.rfp_backbone.get('pretrained'):
                cfg.model.neck.rfp_backbone.pretrained = None

    cfg.gpu_ids = [args.gpu_id]
    cfg.device = get_device()

    # 配置测试数据集模式
    cfg.data.test.test_mode = True
    if cfg.data.test_dataloader.get('samples_per_gpu', 1) > 1:
        # 将 “ImageToTensor ”替换为 “DefaultFormatBundle”。
        cfg.data.test.pipeline = replace_ImageToTensor(cfg.data.test.pipeline)

    print(f"Test dataset config: {cfg.data.test}")
    # 构建数据加载器
    dataset = build_dataset(cfg.data.test)
    data_loader = build_dataloader(dataset, samples_per_gpu=1, workers_per_gpu=2, shuffle=False)
    # 确认数据集是否加载成功
    print(f"Total samples in dataset: {len(dataset)}")
    print(f"Data infos: {dataset.data_infos}")

    # 建立模型并加载检查点
    cfg.model.train_cfg = None
    model = build_detector(cfg.model, test_cfg=cfg.get('test_cfg'))
    checkpoint = load_checkpoint(model, args.checkpoint, map_location='cpu',strict=False)

    # 使用 GPU 进行推理
    model = build_dp(model, cfg.device, device_ids=cfg.gpu_ids)

    # 保存分割结果
    if args.show_dir:
        mmcv.mkdir_or_exist(args.show_dir)

    outputs = single_gpu_test(model, data_loader, show=False)

    # 遍历输出，保存分割结果到指定目录
    for i, result in enumerate(outputs):
        img_info = dataset.img_infos[i]
        img_path = osp.join(dataset.img_prefix, img_info['filename'])
        img = mmcv.imread(img_path)
        result_img = model.show_result(img, result, score_thr=args.show_score_thr, show=False)

        save_path = osp.join(args.show_dir, osp.basename(img_path))
        mmcv.imwrite(result_img, save_path)
        print(f'Saved segmentation result to {save_path}')

if __name__ == '__main__':
    main()
以上是我新建的测试代码...
_base_ = [
    '../_base_/models/mask_rcnn_r50_fpn.py',
    '../_base_/schedules/schedule_1x.py',
    '../_base_/default_runtime.py'
]
dataset_type = 'CustomDataset'
data_root = 'E:/Model/WaterMask-master/new_dataset/'
img_norm_cfg = dict(
    mean = [81.236, 113.761, 117.095], std = [60.598, 58.471, 62.821], to_rgb=True)

model = dict(
    roi_head=dict(
        type='WaterRoIHead',
        mask_head=dict(
            _delete_=True,
            type='WaterMaskHead',
            num_convs_gff=2,
            num_convs_lcf=2,
            image_patch_token=3,
            graph_top_k=11,
            num_heads_in_gat=1,
            classes_num_in_stages=[7, 7, 1],
            stage_output_mask_size=[14, 28, 56],
            loss_cfg=dict(
                type='LaplacianCrossEntropyLoss',
                stage_lcf_loss_weight=[0.25, 0.65, 1],
                boundary_width=3,
                start_stage=2)
        )
    )
)

test_pipeline=[
    dict(type='LoadImageFromFile'),  # 从文件加载图像
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1333, 800),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),  # 调整图像大小
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),  # 填充图像到32的倍数
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img'])  # 收集图像键
            ])
]

data = dict(
    test_dataloader=dict(
        samples_per_gpu=1,  # 这里设置了每个 GPU 处理 2 张图像
        workers_per_gpu=2   # 这里设置了每个 GPU 有 2 个 worker 线程
    ),
    test=dict(
        img_prefix=data_root + 'Images/',
        type=dataset_type,
        ann_file=None,
#        ann_file=data_root + 'annotations/val.json',
        pipeline=test_pipeline
    )
)

optimizer = dict(type='SGD', lr=0.0025, momentum=0.9, weight_decay=0.0001)
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=0.001,
    step=[8,11])
optimizer_config = dict(_delete_=True, grad_clip=dict(max_norm=35, norm_type=2))

evaluation = None
checkpoint_config = dict(interval=3)
以上是我的配置文件，为什么我跳不过标注文件部分？
![image](https://github.com/user-attachments/assets/92bb659a-a0ac-4979-8368-206cb9132e77)
我已经修改了custom类，但还是会报出没有标注文件的错误，怎么跳过标注文件直接对图像进行检测分割呢？

**Comment by LiamLian0727** [comment]
If you use the test.py script, you must provide a COCO-format json file, but the json file do not need to contain annotations, only information such as image paths. If you want to directly call the model inference, you can refer to the [this](https://github.com/open-mmlab/mmdetection/blob/v2.25.1/demo/inference_demo.ipynb)
